src/intrinsic/mod.rs

```diff
@@ -967,35 +967,64 @@ impl<'a, 'gcc, 'tcx> Builder<'a, 'gcc, 'tcx> {
     }
 
     fn saturating_add(&mut self, lhs: RValue<'gcc>, rhs: RValue<'gcc>, signed: bool, width: u64) -> RValue<'gcc> {
-        let func = self.current_func.borrow().expect("func");
+        let result_type = lhs.get_type();
+        let supports_native_type = self.is_native_int_type(result_type);
 
-        if signed {
-            // Algorithm from: https://stackoverflow.com/a/56531252/389119
-            let after_block = func.new_block("after");
-            let func_name =
-                match width {
-                    8 => "__builtin_add_overflow",
-                    16 => "__builtin_add_overflow",
-                    32 => "__builtin_sadd_overflow",
-                    64 => "__builtin_saddll_overflow",
-                    128 => "__builtin_add_overflow",
-                    _ => unreachable!(),
-                };
-            let overflow_func = self.context.get_builtin_function(func_name);
-            let result_type = lhs.get_type();
+        if signed || !supports_native_type {
+            // Based on algorithm from: https://stackoverflow.com/a/56531252/389119
+            let func = self.current_func.borrow().expect("func");
             let res = func.new_local(None, result_type, "saturating_sum");
-            let overflow = self.overflow_call(overflow_func, &[lhs, rhs, res.get_address(None)], None);
+            let overflow =
+                if supports_native_type {
+                    let func_name =
+                        match (width, signed) {
+                            (8, true) => "__builtin_add_overflow",
+                            (16, true) => "__builtin_add_overflow",
+                            (32, true) => "__builtin_sadd_overflow",
+                            (64, true) => "__builtin_saddll_overflow",
+                            (128, true) => "__builtin_add_overflow",
+                            _ => unreachable!(),
+                        };
+                    let overflow_func = self.context.get_builtin_function(func_name);
+                    self.overflow_call(overflow_func, &[lhs, rhs, res.get_address(None)], None)
+                }
+                else {
+                    let func_name =
+                        match (width, signed) {
+                            (128, true) => "__rust_i128_addo",
+                            (128, false) => "__rust_u128_addo",
```

Comment:
I was planning to remove these from compiler-builtins eventually. They should be implementable using just a couple of instructions.

Reply:
I see. However, this family of functions is already in use in two places (`gcc_checked_binop` and `saturating_sub`), and it feels inconsistent to change this in only one place, Maybe this should be done in a separate patchset?

https://github.com/rust-lang/rustc_codegen_gcc/blob/341b9f274bddcb912a6109c57ee236f3c9529649/src/int.rs#L301-L306

https://github.com/rust-lang/rustc_codegen_gcc/blob/341b9f274bddcb912a6109c57ee236f3c9529649/src/intrinsic/mod.rs#L1043

Reply:
I see. Let's leave the `__rust_i128_addo` call for now.